File: 이재표/과자 나눠주기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int[] in = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] cookie = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int M = in[0];
        int N = in[1];
        Arrays.sort(cookie);
        int start = 1;
        int end = cookie[cookie.length - 1];
        while (start <= end) {
            int mid = (start + end) / 2;
            int answer = 0;
            for (int i = 0; i < N; i++) {
                answer += cookie[i] / mid;
            }
            if (answer < M) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        System.out.println(end);
    }
}

File: 이연걸/과자나눠주기.py
# 과자 나눠주기
m, n = map(int, input().split())
snacks = list(map(int, input().split()))
_max = max(snacks)

'''
6 3
10 10 15
정답: 5

10 3
3 3 3
정답: 0

1 1
100
정답: 100

10 10
100 100 100 100 100 100 100 100 100 100 100
정답: 100

10 10
1 1 1 1 1 1 1 1 1 1 1
정답: 1
'''

answer = 0
st, en = 1, _max
while st <= en:
    mid = (st + en) // 2
    if mid == 0: break

    snacks_cnt = 0
    for s in snacks: snacks_cnt += s // mid

    if snacks_cnt < m:
        en = mid - 1
    elif snacks_cnt >= m:
        answer = max(answer, mid)
        st = mid + 1
print(answer)
